# Dealing with Different Datasets

While working on problems in analytics, we generally have to deal with the following two aspects,
- Multiple data sources (as different datasets)
- Multiple data types and formats for the same piece of data

This requires us to combine multiple datasets to create a single holistic dataset for further analysis and model building.

## Combining Datasets

Pandas is rich in its support for combining datasets and data with different datatypes. It also provides functionality to support set logic for indexes and relational algebra.

### `concat()` Function

The `concat()` function does the underlying heavy lifting of performing concatenation operations, for two or more dataframes, along an axis. It also allows performing of optional set operations (e.g. union or intersection) of the indexes (if any) on the other axes.

In [7]:
# Consider the simple example of concatenating 3 dataframes as shown below

In [1]:
import pandas as pd
import numpy as np

In [6]:
df1 = pd.read_csv("Data_set/sample_1.csv", index_col="ID")

In [7]:
df1

A   B   C   D
ID                
0   A0  B0  C0  D0
1   A1  B1  C1  D1
2   A2  B2  C2  D2
3   A3  B3  C3  D3

In [8]:
df2 = pd.read_csv("Data_set/Sample_2.csv")
df2

ID   A   B   C   D
0   4  A4  B4  C4  D4
1   5  A5  B5  C5  D5
2   6  A6  B6  C6  D6
3   7  A7  B7  C7  D7

In [9]:
df3 = pd.read_csv("Data_set/Sample_3.csv")
df3

ID    A    B    C    D
0   8   A8   B8   C8   D8
1   9   A9   B9   C9   D9
2  10  A10  B10  C10  D10
3  11  A11  B11  C11  D11

In [10]:
frames = [df1, df2, df3]   # create a list of dataframes

In [16]:
result1 = pd.concat(frames, axis=1)

In [19]:
result2 = pd.concat(frames, axis=0)

In [20]:
print(result1)
print("-"*40)
print(result2)

    A   B   C   D  ID   A   B   C   D  ID    A    B    C    D
0  A0  B0  C0  D0   4  A4  B4  C4  D4   8   A8   B8   C8   D8
1  A1  B1  C1  D1   5  A5  B5  C5  D5   9   A9   B9   C9   D9
2  A2  B2  C2  D2   6  A6  B6  C6  D6  10  A10  B10  C10  D10
3  A3  B3  C3  D3   7  A7  B7  C7  D7  11  A11  B11  C11  D11
----------------------------------------
     A    B    C    D    ID
0   A0   B0   C0   D0   NaN
1   A1   B1   C1   D1   NaN
2   A2   B2   C2   D2   NaN
3   A3   B3   C3   D3   NaN
0   A4   B4   C4   D4   4.0
1   A5   B5   C5   D5   5.0
2   A6   B6   C6   D6   6.0
3   A7   B7   C7   D7   7.0
0   A8   B8   C8   D8   8.0
1   A9   B9   C9   D9   9.0
2  A10  B10  C10  D10  10.0
3  A11  B11  C11  D11  11.0


`pandas.concat()` takes a list or dict of homogeneously-typed objects and concatenates them with some configurable handling of “what to do with the other axes”

In [16]:
# Syntax
# pd.concat(objs, axis=0, join='outer', ignore_index=False, keys=None,
#          levels=None, names=None, verify_integrity=False, copy=True)

In [17]:
help(pd.concat)   # Go through the help to understand various parameters

Help on function concat in module pandas.core.reshape.concat:

concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False, keys=None, levels=None, names=None, verify_integrity=False, sort=None, copy=True)
    Concatenate pandas objects along a particular axis with optional set logic
    along the other axes.
    
    Can also add a layer of hierarchical indexing on the concatenation axis,
    which may be useful if the labels are the same (or overlapping) on
    the passed axis number.
    
    Parameters
    ----------
    objs : a sequence or mapping of Series or DataFrame objects
        If a dict is passed, the sorted keys will be used as the `keys`
        argument, unless it is passed, in which case the values will be
        selected (see below). Any None objects will be dropped silently unless
        they are all None in which case a ValueError will be raised.
    axis : {0/'index', 1/'columns'}, default 0
        The axis to concatenate along.
    join : {'inner', 

**Example: Index data for each dataframe after concatenation**

In [25]:
result1 = pd.concat(frames, keys=['df1', 'df2', 'df3'], axis=1)  # We add a key, one for each dataframe. df1 for sample_1.csv and so on.

In [28]:
result = pd.concat(frames, keys=['df1', 'df2', 'df3'])  # We add a key, one for each dataframe. df1 for sample_1.csv and so on.

In [26]:
result2 = pd.concat(frames, keys=['df1', 'df2', 'df3'], axis=0)  # We add a key, one for each dataframe. df1 for sample_1.csv and so on.

In [27]:
print(result1)
print("-"*40)
print(result2)

  df1             df2                 df3                    
    A   B   C   D  ID   A   B   C   D  ID    A    B    C    D
0  A0  B0  C0  D0   4  A4  B4  C4  D4   8   A8   B8   C8   D8
1  A1  B1  C1  D1   5  A5  B5  C5  D5   9   A9   B9   C9   D9
2  A2  B2  C2  D2   6  A6  B6  C6  D6  10  A10  B10  C10  D10
3  A3  B3  C3  D3   7  A7  B7  C7  D7  11  A11  B11  C11  D11
----------------------------------------
          A    B    C    D    ID
    ID                          
df1 0    A0   B0   C0   D0   NaN
    1    A1   B1   C1   D1   NaN
    2    A2   B2   C2   D2   NaN
    3    A3   B3   C3   D3   NaN
df2 0    A4   B4   C4   D4   4.0
    1    A5   B5   C5   D5   5.0
    2    A6   B6   C6   D6   6.0
    3    A7   B7   C7   D7   7.0
df3 0    A8   B8   C8   D8   8.0
    1    A9   B9   C9   D9   9.0
    2   A10  B10  C10  D10  10.0
    3   A11  B11  C11  D11  11.0


We get a hierarchical index, and we can select a chunk (corresponding to the original dataframe) by key.

In [35]:
result.loc['df3']
# result.loc['df3'][:,'A']

A    B    C    D    ID
ID                          
0    A8   B8   C8   D8   8.0
1    A9   B9   C9   D9   9.0
2   A10  B10  C10  D10  10.0
3   A11  B11  C11  D11  11.0

### Set Operations on Other Axes

The default axis for concatenation is `axis=0` (i.e. rows will be added). `axis=1` represents columns. While performing concatenation on one axis we can specify set operation on the other axis. 

The default set operation is `join=outer`, i.e. take union of all the datasets. The other set operation is `join=inner`, corresponding to an intersection of the datasets.

In [36]:
# Example: Union
df4 = pd.read_csv("Data_set/Sample_4.csv", index_col="ID")

In [37]:
df1

A   B   C   D
ID                
0   A0  B0  C0  D0
1   A1  B1  C1  D1
2   A2  B2  C2  D2
3   A3  B3  C3  D3

In [38]:
df4

B   D   F
ID            
2   B2  D2  F2
3   B3  D3  F3
6   B6  D6  F6
7   B7  D7  F7

In [53]:
result = pd.concat([df1, df4], axis=1, sort=False)  # Default is outer join. We are doing an outer join on axis=1

In [54]:
result    # Notice how NaN value is added to the result dataframe

A    B    C    D    B    D    F
ID                                   
0    A0   B0   C0   D0  NaN  NaN  NaN
1    A1   B1   C1   D1  NaN  NaN  NaN
2    A2   B2   C2   D2   B2   D2   F2
3    A3   B3   C3   D3   B3   D3   F3
6   NaN  NaN  NaN  NaN   B6   D6   F6
7   NaN  NaN  NaN  NaN   B7   D7   F7

In [59]:
# Example: Intersection
result = pd.concat([df1, df4], axis=1, join='inner')

In [60]:
result

A   B   C   D   B   D   F
ID                            
2   A2  B2  C2  D2  B2  D2  F2
3   A3  B3  C3  D3  B3  D3  F3

### `append()` Instance Method

A useful shortcut to `concat()` is the `append()` instance methods on Series and DataFrame. These methods actually pre-dated `concat()`. They concatenate along axis=0, namely the index.

In [61]:
result = df1.append(df2)

In [63]:
result

A   B   C   D   ID
0  A0  B0  C0  D0  NaN
1  A1  B1  C1  D1  NaN
2  A2  B2  C2  D2  NaN
3  A3  B3  C3  D3  NaN
0  A4  B4  C4  D4  4.0
1  A5  B5  C5  D5  5.0
2  A6  B6  C6  D6  6.0
3  A7  B7  C7  D7  7.0

In [64]:
# Concatenate multiple dataframes at once
result = df1.append([df2, df3])

In [65]:
result

A    B    C    D    ID
0   A0   B0   C0   D0   NaN
1   A1   B1   C1   D1   NaN
2   A2   B2   C2   D2   NaN
3   A3   B3   C3   D3   NaN
0   A4   B4   C4   D4   4.0
1   A5   B5   C5   D5   5.0
2   A6   B6   C6   D6   6.0
3   A7   B7   C7   D7   7.0
0   A8   B8   C8   D8   8.0
1   A9   B9   C9   D9   9.0
2  A10  B10  C10  D10  10.0
3  A11  B11  C11  D11  11.0

In [68]:
# If datasets have overlapping indexes then we can ignore the overlap and concat all the rows
result = pd.concat([df1, df4], ignore_index=True, sort=False)
result

A   B    C   D    F
0   A0  B0   C0  D0  NaN
1   A1  B1   C1  D1  NaN
2   A2  B2   C2  D2  NaN
3   A3  B3   C3  D3  NaN
4  NaN  B2  NaN  D2   F2
5  NaN  B3  NaN  D3   F3
6  NaN  B6  NaN  D6   F6
7  NaN  B7  NaN  D7   F7

In [33]:
# TRY OUT: Set ignore_index=False, and compare the output

In [69]:
# This is also true for append
result = df1.append(df4, ignore_index=True, sort=False)

In [70]:
result

A   B    C   D    F
0   A0  B0   C0  D0  NaN
1   A1  B1   C1  D1  NaN
2   A2  B2   C2  D2  NaN
3   A3  B3   C3  D3  NaN
4  NaN  B2  NaN  D2   F2
5  NaN  B3  NaN  D3   F3
6  NaN  B6  NaN  D6   F6
7  NaN  B7  NaN  D7   F7

### Concatenate Mixed Objects: Series and DataFrame

`concat()` allows us to concatenate a combination of Series and DataFrame objects. The Series is converted to a DataFrame with the column name as the name of the Series.

In [89]:
s1 = pd.Series(['Z0', 'Z1', 'Z2', 'Z3'], name='Z')
s1

0    Z0
1    Z1
2    Z2
3    Z3
Name: Z, dtype: object

In [72]:
result = pd.concat([df1, s1], axis=1)

In [73]:
result

A   B   C   D   Z
0  A0  B0  C0  D0  Z0
1  A1  B1  C1  D1  Z1
2  A2  B2  C2  D2  Z2
3  A3  B3  C3  D3  Z3

In [74]:
# Unnamed Series, on concatenation will be numbered consecutively.
s2 = pd.Series(['_0', '_1', '_2', '_3'])
s2

0    _0
1    _1
2    _2
3    _3
dtype: object

In [83]:
result = pd.concat([df1, s2, s2, s2], axis=1)

In [84]:
result

A   B   C   D   0   1   2
0  A0  B0  C0  D0  _0  _0  _0
1  A1  B1  C1  D1  _1  _1  _1
2  A2  B2  C2  D2  _2  _2  _2
3  A3  B3  C3  D3  _3  _3  _3

<br>

In [85]:
# Passing ignore_index=True will drop all name references. See the effect on column names.
result = pd.concat([df1, s1], axis=1, ignore_index=True)
result

0   1   2   3   4
0  A0  B0  C0  D0  Z0
1  A1  B1  C1  D1  Z1
2  A2  B2  C2  D2  Z2
3  A3  B3  C3  D3  Z3

In [95]:
result = df1.append(s1, ignore_index=True)
result

A    B    C    D    0    1    2    3
0   A0   B0   C0   D0  NaN  NaN  NaN  NaN
1   A1   B1   C1   D1  NaN  NaN  NaN  NaN
2   A2   B2   C2   D2  NaN  NaN  NaN  NaN
3   A3   B3   C3   D3  NaN  NaN  NaN  NaN
4  NaN  NaN  NaN  NaN   Z0   Z1   Z2   Z3

## Appending Rows to a DataFrame

We can append a single row to a DataFrame by passing a Series or dict to `append()`, which returns a new DataFrame. A new object is created, hence the method is not particularly efficient.

In [91]:
import pandas as pd

In [92]:
s2 = pd.Series(['X0', 'X1', 'X2', 'X3'], index=['A', 'B', 'C', 'D'])
s2

A    X0
B    X1
C    X2
D    X3
dtype: object

In [96]:
df1

A   B   C   D
ID                
0   A0  B0  C0  D0
1   A1  B1  C1  D1
2   A2  B2  C2  D2
3   A3  B3  C3  D3

In [93]:
result = df1.append(s2, ignore_index=True)  # Since the series is not labeled

In [94]:
result

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3
4  X0  X1  X2  X3

In [97]:
# Example: Append with a list of dicts or Series
dicts = [{'A': 1, 'B': 2, 'C': 3, 'X': 4}, {'A': 5, 'B': 6, 'C': 7, 'Y': 8}]
dicts

[{'A': 1, 'B': 2, 'C': 3, 'X': 4}, {'A': 5, 'B': 6, 'C': 7, 'Y': 8}]

In [98]:
result = df1.append(dicts, ignore_index=True, sort=False)

In [99]:
result

A   B   C    D    X    Y
0  A0  B0  C0   D0  NaN  NaN
1  A1  B1  C1   D1  NaN  NaN
2  A2  B2  C2   D2  NaN  NaN
3  A3  B3  C3   D3  NaN  NaN
4   1   2   3  NaN  4.0  NaN
5   5   6   7  NaN  NaN  8.0

## Database-style Joining with `merge()` and `join()`

pandas has full-featured, high performance in-memory join operations, very similar to the relational databases like SQL. pandas provides a single function (and an instance method on DataFrame), `merge()`, as the entry point for all standard database join operations between DataFrame or named Series objects.

In [50]:
# Syntax
# pd.merge(left, right, how='inner', on=None, left_on=None, right_on=None,
#          left_index=False, right_index=False, sort=True,
#          suffixes=('_x', '_y'), copy=True, indicator=False,
#          validate=None)

In [51]:
help(pd.merge)    # Read the documentation to understand each parameter in detail

Help on function merge in module pandas.core.reshape.merge:

merge(left, right, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, sort=False, suffixes=('_x', '_y'), copy=True, indicator=False, validate=None)
    Merge DataFrame or named Series objects with a database-style join.
    
    The join is done on columns or indexes. If joining columns on
    columns, the DataFrame indexes *will be ignored*. Otherwise if joining indexes
    on indexes or indexes on a column or columns, the index will be passed on.
    
    Parameters
    ----------
    left : DataFrame
    right : DataFrame or named Series
        Object to merge with.
    how : {'left', 'right', 'outer', 'inner'}, default 'inner'
        Type of merge to be performed.
    
        * left: use only keys from left frame, similar to a SQL left outer join;
          preserve key order.
        * right: use only keys from right frame, similar to a SQL right outer join;
          preserve key 

There is another related method `join()`, that uses `merge()` internally for the index-on-index (by default) and column(s)-on-index join.

### Relational Algebra with `merge()`

Please revise the different join types (which were done during the MySQL course) such as,
- one-to-one
- one-to-many
- many-to-one
- many-to-many

Further review, **left** and **right** join types.

In [100]:
# Simple example with Unique Key combinations
left = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3'], 
                     'A': ['A0', 'A1', 'A2', 'A3'],
                     'B': ['B0', 'B1', 'B2', 'B3']})
left

key   A   B
0  K0  A0  B0
1  K1  A1  B1
2  K2  A2  B2
3  K3  A3  B3

In [101]:
right = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3'],
                      'C': ['C0', 'C1', 'C2', 'C3'],
                      'D': ['D0', 'D1', 'D2', 'D3']})
right

key   C   D
0  K0  C0  D0
1  K1  C1  D1
2  K2  C2  D2
3  K3  C3  D3

In [102]:
# Here we are joining based on the column 'key'
result = pd.merge(left, right, on='key')
result

key   A   B   C   D
0  K0  A0  B0  C0  D0
1  K1  A1  B1  C1  D1
2  K2  A2  B2  C2  D2
3  K3  A3  B3  C3  D3

In [103]:
# Example: Join on multiple keys
left = pd.DataFrame({'key1': ['K0', 'K0', 'K1', 'K2'],
                     'key2': ['K0', 'K1', 'K0', 'K1'],
                     'A': ['A0', 'A1', 'A2', 'A3'],
                     'B': ['B0', 'B1', 'B2', 'B3']})
left

key1 key2   A   B
0   K0   K0  A0  B0
1   K0   K1  A1  B1
2   K1   K0  A2  B2
3   K2   K1  A3  B3

In [104]:
right = pd.DataFrame({'key1': ['K0', 'K1', 'K1', 'K2'],
                      'key2': ['K0', 'K0', 'K0', 'K0'],
                      'C': ['C0', 'C1', 'C2', 'C3'],
                      'D': ['D0', 'D1', 'D2', 'D3']})
right

key1 key2   C   D
0   K0   K0  C0  D0
1   K1   K0  C1  D1
2   K1   K0  C2  D2
3   K2   K0  C3  D3

In [105]:
result = pd.merge(left, right, on=['key1', 'key2'])
result

key1 key2   A   B   C   D
0   K0   K0  A0  B0  C0  D0
1   K1   K0  A2  B2  C1  D1
2   K1   K0  A2  B2  C2  D2

We can specify left or right join types wit the `how` argument to merge. If a key combination does not appear in either the left or right table, the values in the result table will be NA.

Following merge methods are supported by `how` parameter,
- left: Use keys from the left frame only (sql equivalent LEFT OUTER JOIN) 
- right: Use keys from the right frame only (sql equivalent RIGHT OUTER JOIN) 
- outer: Use union of keys from both the frames (sql equivalent FULL OUTER JOIN) 
- inner: Use intersection of keys from both the frames (sql equivalent INNER JOIN) 

In [106]:
result = pd.merge(left, right, how='left', on=['key1', 'key2'])
result

key1 key2   A   B    C    D
0   K0   K0  A0  B0   C0   D0
1   K0   K1  A1  B1  NaN  NaN
2   K1   K0  A2  B2   C1   D1
3   K1   K0  A2  B2   C2   D2
4   K2   K1  A3  B3  NaN  NaN

In [107]:
result = pd.merge(left, right, how='right', on=['key1', 'key2'])
result

key1 key2    A    B   C   D
0   K0   K0   A0   B0  C0  D0
1   K1   K0   A2   B2  C1  D1
2   K1   K0   A2   B2  C2  D2
3   K2   K0  NaN  NaN  C3  D3

In [108]:
result = pd.merge(left, right, how='outer', on=['key1', 'key2'])
result

key1 key2    A    B    C    D
0   K0   K0   A0   B0   C0   D0
1   K0   K1   A1   B1  NaN  NaN
2   K1   K0   A2   B2   C1   D1
3   K1   K0   A2   B2   C2   D2
4   K2   K1   A3   B3  NaN  NaN
5   K2   K0  NaN  NaN   C3   D3

In [109]:
result = pd.merge(left, right, how='inner', on=['key1', 'key2'])
result

key1 key2   A   B   C   D
0   K0   K0  A0  B0  C0  D0
1   K1   K0  A2  B2  C1  D1
2   K1   K0  A2  B2  C2  D2

In [110]:
# Example: Join with duplicate keys
left = pd.DataFrame({'A': [1, 2], 'B': [2, 2]})
print(left)
print("-"*40)
right = pd.DataFrame({'A': [4, 5, 6], 'B': [2, 2, 2]})
print(right)
print("-"*40)
result = pd.merge(left, right, on='B', how='outer')
print(result)

   A  B
0  1  2
1  2  2
----------------------------------------
   A  B
0  4  2
1  5  2
2  6  2
----------------------------------------
   A_x  B  A_y
0    1  2    4
1    1  2    5
2    1  2    6
3    2  2    4
4    2  2    5
5    2  2    6


Presence of data with duplicate keys, can result in a very large result on merging. Hence, we need to check for duplicats before merging.

### Duplicate Keys

It is always prudent to check for duplicates if you are not suspecting them. With the `validate` argument we can enforce duplicate check before merging is initiated.

Consider the following example. There are duplicate values of B in the right DataFrame. Setting `validate` as 'one_to_one' in this case will  result in an exception on merging.

In [111]:
left = pd.DataFrame({'A' : [1,2], 'B' : [1, 2]})
right = pd.DataFrame({'A' : [4,5,6], 'B': [2, 2, 2]})

print(left)
print("-"*40)
print(right)


   A  B
0  1  1
1  2  2
----------------------------------------
   A  B
0  4  2
1  5  2
2  6  2


In [112]:
# This will throw an error
result = pd.merge(left, right, on='B', how='outer', validate="one_to_one")

MergeError: Merge keys are not unique in right dataset; not a one-to-one merge

In [114]:
# Example: Allow duplicates in the right but NOT in the left
pd.merge(left, right, on='B', how='outer', validate="one_to_many")

A_x  B  A_y
0    1  1  NaN
1    2  2  4.0
2    2  2  5.0
3    2  2  6.0

### The Merge Indicator

`merge()` accepts an argument called `indicator`. If set to True, a Categorical-type column called `_merge` will be added to the output object that indicates why a row is in the merge output.

In [120]:
df1 = pd.DataFrame({'col1': [0, 1], 'col_left': ['a', 'b']})
print(df1)
print('-'*40)
df2 = pd.DataFrame({'col1': [1, 2, 2], 'col_right': [2, 2, 2]})
print(df2)
print('-'*40)
print(pd.merge(df1, df2, on='col1', how='outer', indicator=True))
print('-'*40)
print(pd.merge(df1, df2, on='col1', how='inner', indicator=True))
print('-'*40)
print(pd.merge(df1, df2, on='col1', how='left', indicator=True))
print('-'*40)
print(pd.merge(df1, df2, on='col1', how='right', indicator=True))


# NOTE: We can rename the _merge column name by passing the column name to "indicator" parameter.
# Please do that.

   col1 col_left
0     0        a
1     1        b
----------------------------------------
   col1  col_right
0     1          2
1     2          2
2     2          2
----------------------------------------
   col1 col_left  col_right      _merge
0     0        a        NaN   left_only
1     1        b        2.0        both
2     2      NaN        2.0  right_only
3     2      NaN        2.0  right_only
----------------------------------------
   col1 col_left  col_right _merge
0     1        b          2   both
----------------------------------------
   col1 col_left  col_right     _merge
0     0        a        NaN  left_only
1     1        b        2.0       both
----------------------------------------
   col1 col_left  col_right      _merge
0     1        b          2        both
1     2      NaN          2  right_only
2     2      NaN          2  right_only


### Datatype Preservation with `merge()`

Merging will preserve the dtype of the join keys, other columns. Even categorical datatype is preserved. In the presence of missing values, the resulting type will be upcast.

In [121]:
import pandas as pd
left = pd.DataFrame({'key': [1], 'v1': [10]})
print("Left\n", left, '\n')
print("-"*40)
right = pd.DataFrame({'key': [1, 2], 'v1': [20, 30]})
print("Right\n", right, '\n')
print("-"*40)
merged = pd.merge(left, right, how='outer')
print("Merged\n", merged, '\n')

Left
    key  v1
0    1  10 

----------------------------------------
Right
    key  v1
0    1  20
1    2  30 

----------------------------------------
Merged
    key  v1
0    1  10
1    1  20
2    2  30 



In [221]:
print("Left data types\n", left.dtypes, '\n')
print("-"*40)
print("Right data types\n", right.dtypes, '\n')
print("-"*40)
print("Merged data types\n", merged.dtypes, '\n')

Left data types
 key    int64
v1     int64
dtype: object 

----------------------------------------
Right data types
 key    int64
v1     int64
dtype: object 

----------------------------------------
Merged data types
 key    int64
v1     int64
dtype: object 



In [122]:
# With missing values, the resulting datatype is upcast from int to float
merged = pd.merge(left, right, how='outer', on='key')
print("Merged\n", merged, '\n')
print("-"*40)
print("Merged data types\n", merged.dtypes, '\n')

Merged
    key  v1_x  v1_y
0    1  10.0    20
1    2   NaN    30 

----------------------------------------
Merged data types
 key       int64
v1_x    float64
v1_y      int64
dtype: object 



In [123]:
# Example of categorical datatype preservation
import numpy as np

In [128]:
from pandas.api.types import CategoricalDtype
X = pd.Series(np.random.choice(['foo', 'bar', 'xyz'], size=(10,)))
print(X)
print("--"*40)
X = X.astype(CategoricalDtype(categories=['foo', 'bar', 'xyz']))
print(X)

0    xyz
1    bar
2    xyz
3    xyz
4    bar
5    foo
6    bar
7    bar
8    xyz
9    bar
dtype: object
--------------------------------------------------------------------------------
0    xyz
1    bar
2    xyz
3    xyz
4    bar
5    foo
6    bar
7    bar
8    xyz
9    bar
dtype: category
Categories (3, object): [foo, bar, xyz]


In [129]:
left = pd.DataFrame({'X': X, 'Y': np.random.choice(['one', 'two', 'three'], size=(10,))})
left

X      Y
0  xyz    two
1  bar    two
2  xyz    two
3  xyz    two
4  bar  three
5  foo  three
6  bar    one
7  bar    two
8  xyz    two
9  bar    two

In [130]:
left.dtypes

X    category
Y      object
dtype: object

In [131]:
right = pd.DataFrame({'X': pd.Series(['foo', 'bar'], dtype=CategoricalDtype(['foo', 'bar'])), 'Z': [1, 2]})

In [132]:
right

X  Z
0  foo  1
1  bar  2

In [76]:
right.dtypes

X    category
Z       int64
dtype: object

In [77]:
result = pd.merge(left, right, how='outer')

In [78]:
result

X      Y  Z
0  bar  three  2
1  bar    two  2
2  bar    one  2
3  bar    two  2
4  bar    two  2
5  bar  three  2
6  foo    two  1
7  foo    one  1
8  foo  three  1
9  foo    two  1

In [79]:
result.dtypes

X    category
Y      object
Z       int64
dtype: object

### Join: Differently Indexed DataFrames

`join()` method supports combining the columns of two potentially differently-indexed DataFrames into a single DataFrame

In [133]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2'], 'B': ['B0', 'B1', 'B2']}, index=['K0', 'K1', 'K2'])
left

A   B
K0  A0  B0
K1  A1  B1
K2  A2  B2

In [134]:
right = pd.DataFrame({'C': ['C0', 'C2', 'C3'], 'D': ['D0', 'D2', 'D3']}, index=['K0', 'K2', 'K3'])
right

C   D
K0  C0  D0
K2  C2  D2
K3  C3  D3

In [135]:
result = left.join(right)    # Left join

In [136]:
result

A   B    C    D
K0  A0  B0   C0   D0
K1  A1  B1  NaN  NaN
K2  A2  B2   C2   D2

In [8]:
result = left.join(right, how='outer')

In [9]:
result

A    B    C    D
K0   A0   B0   C0   D0
K1   A1   B1  NaN  NaN
K2   A2   B2   C2   D2
K3  NaN  NaN   C3   D3

In [86]:
result = left.join(right, how='inner')

In [87]:
result

A   B   C   D
K0  A0  B0  C0  D0
K2  A2  B2  C2  D2

In the above joins, the merging is based on the indexes (row labels). With `merge()`, the same effect is achieved by explicitly specifying `merge()` to use indexes, as shown below.

In [88]:
result = pd.merge(left, right, left_index=True, right_index=True, how='outer')

In [89]:
result

A    B    C    D
K0   A0   B0   C0   D0
K1   A1   B1  NaN  NaN
K2   A2   B2   C2   D2
K3  NaN  NaN   C3   D3

In [90]:
result = pd.merge(left, right, left_index=True, right_index=True, how='inner')

In [91]:
result

A   B   C   D
K0  A0  B0  C0  D0
K2  A2  B2  C2  D2

### Join: Key Columns on an Index

With `join()` we can specify one or more column(s) on which to align merge operations. Similarly with `merge()` we can achieve the same effect, as shown below.

In [92]:
# Syntax
# left.join(right, on=key_or_keys)
# pd.merge(left, right, left_on=key_or_keys, right_index=True, how='left', sort=False)

Obviously user can choose whichever form find more convenient. For many-to-one joins (where one of the DataFrame’s is already indexed by the join key), using join may be more convenient. Here is a simple example:

In [137]:
# Example: many-to-one join, more convenient to use join than merge in this case
left = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'], 'B': ['B0', 'B1', 'B2', 'B3'], 'key': ['K0', 'K1', 'K0', 'K1']})
left

A   B key
0  A0  B0  K0
1  A1  B1  K1
2  A2  B2  K0
3  A3  B3  K1

In [139]:
right = pd.DataFrame({'C': ['C0', 'C1'], 'D': ['D0', 'D1']}, index=['K0', 'K1'])
right

C   D
K0  C0  D0
K1  C1  D1

In [95]:
result = left.join(right, on='key')

In [96]:
result

A   B key   C   D
0  A0  B0  K0  C0  D0
1  A1  B1  K1  C1  D1
2  A2  B2  K0  C0  D0
3  A3  B3  K1  C1  D1

In [97]:
result = pd.merge(left, right, left_on='key', right_index=True, how='left', sort=False)

In [98]:
result

A   B key   C   D
0  A0  B0  K0  C0  D0
1  A1  B1  K1  C1  D1
2  A2  B2  K0  C0  D0
3  A3  B3  K1  C1  D1

### Join: Single Index and Multi-Index DataFrames

We can join a single indexed DataFrame with a level of a Multi-Indexed DataFrame. Let's see an example.

In [140]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2'], 'B': ['B0', 'B1', 'B2']}, index=pd.Index(['K0', 'K1', 'K2'], name='key'))
left

A   B
key        
K0   A0  B0
K1   A1  B1
K2   A2  B2

In [141]:
index = pd.MultiIndex.from_tuples([('K0', 'Y0'), ('K1', 'Y1'), ('K2', 'Y2'), ('K2', 'Y3')], names=['key', 'Y'])
index

MultiIndex([('K0', 'Y0'),
            ('K1', 'Y1'),
            ('K2', 'Y2'),
            ('K2', 'Y3')],
           names=['key', 'Y'])

In [142]:
right = pd.DataFrame({'C': ['C0', 'C1', 'C2', 'C3'], 'D': ['D0', 'D1', 'D2', 'D3']}, index=index)
right

C   D
key Y         
K0  Y0  C0  D0
K1  Y1  C1  D1
K2  Y2  C2  D2
    Y3  C3  D3

In [102]:
result = left.join(right, how='inner')# single index 'key' in left dataframe matched with 'key' in multi-index right dataframe

In [103]:
result

A   B   C   D
key Y                 
K0  Y0  A0  B0  C0  D0
K1  Y1  A1  B1  C1  D1
K2  Y2  A2  B2  C2  D2
    Y3  A2  B2  C3  D3

In [104]:
# Achieving same results with merge -- slightly longer to type
result = pd.merge(left.reset_index(), right.reset_index(), on=['key'], how='inner').set_index(['key','Y'])

In [143]:
print(left.reset_index())
print("-"*40)
print(right.reset_index())

  key   A   B
0  K0  A0  B0
1  K1  A1  B1
2  K2  A2  B2
----------------------------------------
  key   Y   C   D
0  K0  Y0  C0  D0
1  K1  Y1  C1  D1
2  K2  Y2  C2  D2
3  K2  Y3  C3  D3


In [105]:
result

A   B   C   D
key Y                 
K0  Y0  A0  B0  C0  D0
K1  Y1  A1  B1  C1  D1
K2  Y2  A2  B2  C2  D2
    Y3  A2  B2  C3  D3

### Join: Two Multi-Indexes

There is limited support for joining two multi-indexed dataframes. Joining is possible only if indices for the right argument are a subset of the indices for the left argument.

In [15]:
leftindex = pd.MultiIndex.from_product([list('abc'), list('xy'), [1, 2]], names=['abc', 'xy', 'num'])
leftindex

MultiIndex([('a', 'x', 1),
            ('a', 'x', 2),
            ('a', 'y', 1),
            ('a', 'y', 2),
            ('b', 'x', 1),
            ('b', 'x', 2),
            ('b', 'y', 1),
            ('b', 'y', 2),
            ('c', 'x', 1),
            ('c', 'x', 2),
            ('c', 'y', 1),
            ('c', 'y', 2)],
           names=['abc', 'xy', 'num'])

In [107]:
left = pd.DataFrame({'v1': range(12)}, index=leftindex)

In [108]:
left

v1
abc xy num    
a   x  1     0
       2     1
    y  1     2
       2     3
b   x  1     4
       2     5
    y  1     6
       2     7
c   x  1     8
       2     9
    y  1    10
       2    11

In [16]:
rightindex = pd.MultiIndex.from_product([list('abc'), list('xy')], names=['abc', 'xy'])
rightindex

MultiIndex([('a', 'x'),
            ('a', 'y'),
            ('b', 'x'),
            ('b', 'y'),
            ('c', 'x'),
            ('c', 'y')],
           names=['abc', 'xy'])

In [110]:
right = pd.DataFrame({'v2': [100 * i for i in range(1, 7)]}, index=rightindex)

In [111]:
right

v2
abc xy     
a   x   100
    y   200
b   x   300
    y   400
c   x   500
    y   600

In [112]:
left.join(right, on=['abc', 'xy'], how='inner')

v1   v2
abc xy num         
a   x  1     0  100
       2     1  100
    y  1     2  200
       2     3  200
b   x  1     4  300
       2     5  300
    y  1     6  400
       2     7  400
c   x  1     8  500
       2     9  500
    y  1    10  600
       2    11  600

If right indices are not a subset of left indices then we can do the merge as follows,

In [113]:
leftindex = pd.MultiIndex.from_tuples([('K0', 'X0'), ('K0', 'X1'), ('K1', 'X2')], names=['key', 'X'])

In [114]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2'], 'B': ['B0', 'B1', 'B2']}, index=leftindex)
left

A   B
key X         
K0  X0  A0  B0
    X1  A1  B1
K1  X2  A2  B2

In [115]:
rightindex = pd.MultiIndex.from_tuples([('K0', 'Y0'), ('K1', 'Y1'), ('K2', 'Y2'), ('K2', 'Y3')], names=['key', 'Y'])

In [116]:
right = pd.DataFrame({'C': ['C0', 'C1', 'C2', 'C3'], 'D': ['D0', 'D1', 'D2', 'D3']}, index=rightindex)
right

C   D
key Y         
K0  Y0  C0  D0
K1  Y1  C1  D1
K2  Y2  C2  D2
    Y3  C3  D3

In [117]:
result = pd.merge(left.reset_index(), right.reset_index(), on=['key'], how='inner').set_index(['key', 'X', 'Y'])

In [118]:
result

A   B   C   D
key X  Y                 
K0  X0 Y0  A0  B0  C0  D0
    X1 Y0  A1  B1  C0  D0
K1  X2 Y1  A2  B2  C1  D1

### Merge: Combination of Columns and Index Levels

The argument values for `on`, `left_on`, and `right_on` parameters could refer to either column names or index level names. Hence we can merge DataFrame instances on a combination of columns and index levels and columns.

In [18]:
left_index = pd.Index(['K0', 'K0', 'K1', 'K2'], name='key1')
left_index

Index(['K0', 'K0', 'K1', 'K2'], dtype='object', name='key1')

In [19]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'], 'B': ['B0', 'B1', 'B2', 'B3'], 
                     'key2': ['K0', 'K1', 'K0', 'K1']}, index=left_index)
left

A   B key2
key1             
K0    A0  B0   K0
K0    A1  B1   K1
K1    A2  B2   K0
K2    A3  B3   K1

In [20]:
right_index = pd.Index(['K0', 'K1', 'K2', 'K2'], name='key1')

In [21]:
right = pd.DataFrame({'C': ['C0', 'C1', 'C2', 'C3'], 'D': ['D0', 'D1', 'D2', 'D3'], 
                      'key2': ['K0', 'K0', 'K0', 'K1']}, index=right_index)
right

C   D key2
key1             
K0    C0  D0   K0
K1    C1  D1   K0
K2    C2  D2   K0
K2    C3  D3   K1

In [22]:
result = left.merge(right, on=['key1', 'key2'])  # match both the index (key1) and the column (key2)

In [124]:
result

A   B key2   C   D
key1                     
K0    A0  B0   K0  C0  D0
K1    A2  B2   K0  C1  D1
K2    A3  B3   K1  C3  D3

### Change Names: Overlapping Columns

The merge suffixes argument takes a tuple of list of strings to append to overlapping column names in the input DataFrames to disambiguate the result columns:

In [145]:
# Example: To disambiguate columns with same name in the merged results, we can add a suffix
left = pd.DataFrame({'k': ['K0', 'K1', 'K2'], 'v': [1, 2, 3]})
left

k  v
0  K0  1
1  K1  2
2  K2  3

In [146]:
right = pd.DataFrame({'k': ['K0', 'K0', 'K3'], 'v': [4, 5, 6]})
right

k  v
0  K0  4
1  K0  5
2  K3  6

In [148]:
result = pd.merge(left, right, on='k', suffixes=['_lefty', '_righty'])
result

k  v_lefty  v_righty
0  K0        1         4
1  K0        1         5

In [149]:
# Similarly, we can do the same with join() using, lsuffix and rsuffix parameters
left = left.set_index('k')
right = right.set_index('k')
result = left.join(right, lsuffix='_l', rsuffix='_r')

In [39]:
result

v_l  v_r
k           
K0    1  4.0
K0    1  5.0
K1    2  NaN
K2    3  NaN

### Join: More than two DataFrames

In [150]:
left

v
k    
K0  1
K1  2
K2  3

In [151]:
right

v
k    
K0  4
K0  5
K3  6

In [152]:
# Example to join three DataFrames!
right2 = pd.DataFrame({'v': [7, 8, 9]}, index=['K1', 'K1', 'K2'])
right2

v
K1  7
K1  8
K2  9

In [43]:
result = left.join([right, right2])

In [44]:
result

v_x  v_y    v
K0    1  4.0  NaN
K0    1  5.0  NaN
K1    2  NaN  7.0
K1    2  NaN  8.0
K2    3  NaN  9.0

### Merge: Values within Series or DataFrame Columns

Consider a scenario, where we have two sensors measuring the same parameter. If due to some fault in the sensors, we  don't get the values always. In such situations, it would be useful to combine the values from different sensors to create a single consolidated observation dataset.

In [153]:
df1 = pd.DataFrame([[np.nan, 3., 5.], [-4.6, np.nan, np.nan], [np.nan, 7., np.nan]]) # data from sensor 1
df1

0    1    2
0  NaN  3.0  5.0
1 -4.6  NaN  NaN
2  NaN  7.0  NaN

In [154]:
df2 = pd.DataFrame([[-42.6, np.nan, -8.2], [-5., 1.6, 4]], index=[1, 2])  # data from sensor 2
df2

0    1    2
1 -42.6  NaN -8.2
2  -5.0  1.6  4.0

For this, use the combine_first() method:

In [155]:
# merge data by taking values from the right DataFrame if they are missing in the left DataFrame
result = df1.combine_first(df2)

In [156]:
result

0    1    2
0  NaN  3.0  5.0
1 -4.6  NaN -8.2
2 -5.0  7.0  4.0

Note that this method only takes values from the right DataFrame if they are missing in the left DataFrame. A related method, update(), alters non-NA values in place:

In [157]:
# update(), alters non-NA values in place
df1.update(df2)
df1

0    1    2
0   NaN  3.0  5.0
1 -42.6  NaN -8.2
2  -5.0  1.6  4.0

### Merge: Ordered Data

`merge_ordered()` function allows combining time series and other ordered data. Perform merge with optional filling/interpolation.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge_ordered.html

In [63]:
help(pd.merge_ordered)

Help on function merge_ordered in module pandas.core.reshape.merge:

merge_ordered(left, right, on=None, left_on=None, right_on=None, left_by=None, right_by=None, fill_method=None, suffixes=('_x', '_y'), how='outer')
    Perform merge with optional filling/interpolation designed for ordered
    data like time series data. Optionally perform group-wise merge (see
    examples)
    
    Parameters
    ----------
    left : DataFrame
    right : DataFrame
    on : label or list
        Field names to join on. Must be found in both DataFrames.
    left_on : label or list, or array-like
        Field names to join on in left DataFrame. Can be a vector or list of
        vectors of the length of the DataFrame to use a particular vector as
        the join key instead of columns
    right_on : label or list, or array-like
        Field names to join on in right DataFrame or vector/list of vectors per
        left_on docs
    left_by : column name or list of column names
        Group left Dat

In [158]:
 left = pd.DataFrame({'k': ['K0', 'K1', 'K1', 'K2'], 'lv': [1, 2, 3, 4], 's': ['a', 'b', 'c', 'd']})
left

k  lv  s
0  K0   1  a
1  K1   2  b
2  K1   3  c
3  K2   4  d

In [159]:
right = pd.DataFrame({'k': ['K1', 'K2', 'K4'], 'rv': [1, 2, 3]})
right

k  rv
0  K1   1
1  K2   2
2  K4   3

In [50]:
# fill_method keyword allows to fill/interpolate missing data
pd.merge_ordered(left, right, left_by='s')

k   lv  s   rv
0   K0  1.0  a  NaN
1   K1  NaN  a  1.0
2   K2  NaN  a  2.0
3   K4  NaN  a  3.0
4   K1  2.0  b  1.0
5   K2  NaN  b  2.0
6   K4  NaN  b  3.0
7   K1  3.0  c  1.0
8   K2  NaN  c  2.0
9   K4  NaN  c  3.0
10  K1  NaN  d  1.0
11  K2  4.0  d  2.0
12  K4  NaN  d  3.0

In [49]:
# fill_method keyword allows to fill/interpolate missing data
pd.merge_ordered(left, right, fill_method='ffill', left_by='s')

k   lv  s   rv
0   K0  1.0  a  NaN
1   K1  1.0  a  1.0
2   K2  1.0  a  2.0
3   K4  1.0  a  3.0
4   K1  2.0  b  1.0
5   K2  2.0  b  2.0
6   K4  2.0  b  3.0
7   K1  3.0  c  1.0
8   K2  3.0  c  2.0
9   K4  3.0  c  3.0
10  K1  NaN  d  1.0
11  K2  4.0  d  2.0
12  K4  4.0  d  3.0

### Nearest Merge with `merge_asof()`

`merge_asof()` facilitates merging of ordered datasets based on NEAREST key rather than EXACT key! Each row in the left DataFrame pairs up with the last row in the right DataFrame whose key value is less than the left’s key. We pass the `on` argument to specify the column name for merging. Optionally we can perform a group-wise merge by specifying the `by` argument. Groupings are based on EXACT matches.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge_asof.html


For example, consider the following DataFrames with trades and quotes. We want to merge them,

In [62]:
help(pd.merge_asof)

Help on function merge_asof in module pandas.core.reshape.merge:

merge_asof(left, right, on=None, left_on=None, right_on=None, left_index=False, right_index=False, by=None, left_by=None, right_by=None, suffixes=('_x', '_y'), tolerance=None, allow_exact_matches=True, direction='backward')
    Perform an asof merge. This is similar to a left-join except that we
    match on nearest key rather than equal keys.
    
    Both DataFrames must be sorted by the key.
    
    For each row in the left DataFrame:
    
      - A "backward" search selects the last row in the right DataFrame whose
        'on' key is less than or equal to the left's key.
    
      - A "forward" search selects the first row in the right DataFrame whose
        'on' key is greater than or equal to the left's key.
    
      - A "nearest" search selects the row in the right DataFrame whose 'on'
        key is closest in absolute distance to the left's key.
    
    The default is "backward" and is compatible in versi

In [53]:
 trades = pd.DataFrame({
        'time': pd.to_datetime(['20160525 13:30:00.023',
                                '20160525 13:30:00.038',
                                '20160525 13:30:00.048',
                                '20160525 13:30:00.048',
                                '20160525 13:30:00.048']),
        'ticker': ['MSFT', 'MSFT',
                   'GOOG', 'GOOG', 'AAPL'],
        'price': [51.95, 51.95,
                  720.77, 720.92, 98.00],
        'quantity': [75, 155,
                     100, 100, 100]},
        columns=['time', 'ticker', 'price', 'quantity'])

In [54]:
quotes = pd.DataFrame({
        'time': pd.to_datetime(['20160525 13:30:00.023',
                                '20160525 13:30:00.023',
                                '20160525 13:30:00.030',
                                '20160525 13:30:00.041',
                                '20160525 13:30:00.048',
                                '20160525 13:30:00.049',
                                '20160525 13:30:00.072',
                                '20160525 13:30:00.075']),
        'ticker': ['GOOG', 'MSFT', 'MSFT',
                   'MSFT', 'GOOG', 'AAPL', 'GOOG',
                   'MSFT'],
        'bid': [720.50, 51.95, 51.97, 51.99,
                720.50, 97.99, 720.50, 52.01],
        'ask': [720.93, 51.96, 51.98, 52.00,
                720.93, 98.01, 720.88, 52.03]},
        columns=['time', 'ticker', 'bid', 'ask']) 

In [55]:
trades

time ticker   price  quantity
0 2016-05-25 13:30:00.023   MSFT   51.95        75
1 2016-05-25 13:30:00.038   MSFT   51.95       155
2 2016-05-25 13:30:00.048   GOOG  720.77       100
3 2016-05-25 13:30:00.048   GOOG  720.92       100
4 2016-05-25 13:30:00.048   AAPL   98.00       100

In [56]:
quotes

time ticker     bid     ask
0 2016-05-25 13:30:00.023   GOOG  720.50  720.93
1 2016-05-25 13:30:00.023   MSFT   51.95   51.96
2 2016-05-25 13:30:00.030   MSFT   51.97   51.98
3 2016-05-25 13:30:00.041   MSFT   51.99   52.00
4 2016-05-25 13:30:00.048   GOOG  720.50  720.93
5 2016-05-25 13:30:00.049   AAPL   97.99   98.01
6 2016-05-25 13:30:00.072   GOOG  720.50  720.88
7 2016-05-25 13:30:00.075   MSFT   52.01   52.03

In [144]:
pd.merge_asof(trades, quotes,
                  on='time',
                  by='ticker')

time ticker   price  quantity     bid     ask
0 2016-05-25 13:30:00.023   MSFT   51.95        75   51.95   51.96
1 2016-05-25 13:30:00.038   MSFT   51.95       155   51.97   51.98
2 2016-05-25 13:30:00.048   GOOG  720.77       100  720.50  720.93
3 2016-05-25 13:30:00.048   GOOG  720.92       100  720.50  720.93
4 2016-05-25 13:30:00.048   AAPL   98.00       100     NaN     NaN

In [145]:
# Define similarity, by specifying the Timedelta as 2ms between quote and trade times
pd.merge_asof(trades, quotes,
                  on='time',
                  by='ticker',
                  tolerance=pd.Timedelta('2ms'))

time ticker   price  quantity     bid     ask
0 2016-05-25 13:30:00.023   MSFT   51.95        75   51.95   51.96
1 2016-05-25 13:30:00.038   MSFT   51.95       155     NaN     NaN
2 2016-05-25 13:30:00.048   GOOG  720.77       100  720.50  720.93
3 2016-05-25 13:30:00.048   GOOG  720.92       100  720.50  720.93
4 2016-05-25 13:30:00.048   AAPL   98.00       100     NaN     NaN

In [146]:
# Specify Timedelta but exclude exact matches!
pd.merge_asof(trades, quotes,
                  on='time',
                  by='ticker',
                  tolerance=pd.Timedelta('10ms'),
                  allow_exact_matches=False)

time ticker   price  quantity    bid    ask
0 2016-05-25 13:30:00.023   MSFT   51.95        75    NaN    NaN
1 2016-05-25 13:30:00.038   MSFT   51.95       155  51.97  51.98
2 2016-05-25 13:30:00.048   GOOG  720.77       100    NaN    NaN
3 2016-05-25 13:30:00.048   GOOG  720.92       100    NaN    NaN
4 2016-05-25 13:30:00.048   AAPL   98.00       100    NaN    NaN

## Pivot Tables

The function `pivot_table()` can be used to create spreadsheet-style pivot tables. The function is also available as an instance method on DataFrame. Let's look at the function parameters.

In [1]:
import pandas as pd
import numpy as np
#help(pd.pivot_table)

In [2]:
import datetime

df = pd.DataFrame({'A': ['one', 'one', 'two', 'three'] * 6,
                   'B': ['A', 'B', 'C'] * 8,
                   'C': ['foo', 'foo', 'foo', 'bar', 'bar', 'bar'] * 4,
                   'D': np.random.randn(24),
                   'E': np.random.randn(24),
                   'F': [datetime.datetime(2013, i, 1) for i in range(1, 13)]+ [datetime.datetime(2013, i, 15) for i in range(1, 13)]})

In [4]:
df

A  B    C         D         E          F
0     one  A  foo  1.028740  0.473510 2013-01-01
1     one  B  foo -0.788913 -0.046557 2013-02-01
2     two  C  foo  0.920739  1.781694 2013-03-01
3   three  A  bar -0.098590  1.155195 2013-04-01
4     one  B  bar  0.437592 -0.814884 2013-05-01
5     one  C  bar -0.413030  0.473212 2013-06-01
6     two  A  foo -0.067134 -0.889710 2013-07-01
7   three  B  foo -0.544514  0.288162 2013-08-01
8     one  C  foo  0.577939 -1.082150 2013-09-01
9     one  A  bar  0.076636  0.434358 2013-10-01
10    two  B  bar  0.625049 -0.980516 2013-11-01
11  three  C  bar  0.067550 -0.410163 2013-12-01
12    one  A  foo  0.158560  0.719794 2013-01-15
13    one  B  foo  1.285111  1.050348 2013-02-15
14    two  C  foo  0.089446 -0.459098 2013-03-15
15  three  A  bar  0.741763 -1.765754 2013-04-15
16    one  B  bar  0.557509  0.746086 2013-05-15
17    one  C  bar -0.125353  1.011254 2013-06-15
18    two  A  foo  0.283445  0.344853 2013-07-15
19  three  B  foo  0.017444 -0.571358 2013-08-15
20    one  C  foo  0.273074 -0.659948 2013-09-15
21    one  A  bar  0.112724 -1.156420 2013-10-15
22    two  B  bar -0.902760  2.179501 2013-11-15
23  three  C  bar -0.009219 -0.679088 2013-12-15

### Export the dataframe and perform the below pivot operations in Excel simulataneously

In [5]:
df.to_csv (r'df_pivot_table.csv', index = None, header=True) #Don't forget to add '.csv' at the end of the path

#### Let's create some pivot tables

In [6]:
# index can be considered as row labels in the output, multiple indexes will be similar to hierarchical indexes
# columns will be the column labels in the output
# values -- aggregate function will apply on the values and result will be output
# default aggregate function is mean

pd.pivot_table(df, values='D', index=['A', 'B'], columns=['C'])

C             bar       foo
A     B                    
one   A  0.094680  0.593650
      B  0.497550  0.248099
      C -0.269192  0.425507
three A  0.321587       NaN
      B       NaN -0.263535
      C  0.029166       NaN
two   A       NaN  0.108155
      B -0.138855       NaN
      C       NaN  0.505092

In [7]:
# single index, multiple columns, with aggregate as sum
pd.pivot_table(df, values='D', index=['B'], columns=['A', 'C'], aggfunc=np.sum)

A       one               three                two          
C       bar       foo       bar      foo       bar       foo
B                                                           
A  0.189360  1.187300  0.643173      NaN       NaN  0.216311
B  0.995101  0.496198       NaN -0.52707 -0.277711       NaN
C -0.538384  0.851013  0.058331      NaN       NaN  1.010185

In [8]:
# more complex, with single index, multiple columns, multiple values!
pd.pivot_table(df, values=['D', 'E'], index=['B'], columns=['A', 'C'], aggfunc=np.sum)

D                                                          E  \
A       one               three                two                 one   
C       bar       foo       bar      foo       bar       foo       bar   
B                                                                        
A  0.189360  1.187300  0.643173      NaN       NaN  0.216311 -0.722062   
B  0.995101  0.496198       NaN -0.52707 -0.277711       NaN -0.068797   
C -0.538384  0.851013  0.058331      NaN       NaN  1.010185  1.484465   

                                                     
A               three                 two            
C       foo       bar       foo       bar       foo  
B                                                    
A  1.193304 -0.610559       NaN       NaN -0.544857  
B  1.003791       NaN -0.283196  1.198985       NaN  
C -1.742097 -1.089251       NaN       NaN  1.322595

In [9]:
# If the values column name is not given, the pivot table will include all of the data 
# that can be aggregated in an additional level of hierarchy in the columns
pd.pivot_table(df, index=['A', 'B'], columns=['C'])

D                   E          
C             bar       foo       bar       foo
A     B                                        
one   A  0.094680  0.593650 -0.361031  0.596652
      B  0.497550  0.248099 -0.034399  0.501895
      C -0.269192  0.425507  0.742233 -0.871049
three A  0.321587       NaN -0.305280       NaN
      B       NaN -0.263535       NaN -0.141598
      C  0.029166       NaN -0.544625       NaN
two   A       NaN  0.108155       NaN -0.272429
      B -0.138855       NaN  0.599492       NaN
      C       NaN  0.505092       NaN  0.661298

### Grouper for Index

In [10]:
# We can have Grouper for index and column keywords
# Below we are indexing on column 'F', datetime with frequency as last date of each month
# See: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Grouper.html
pd.pivot_table(df, values='D', index=pd.Grouper(freq='M', key='F'), columns='C')

C                bar       foo
F                             
2013-01-31       NaN  0.593650
2013-02-28       NaN  0.248099
2013-03-31       NaN  0.505092
2013-04-30  0.321587       NaN
2013-05-31  0.497550       NaN
2013-06-30 -0.269192       NaN
2013-07-31       NaN  0.108155
2013-08-31       NaN -0.263535
2013-09-30       NaN  0.425507
2013-10-31  0.094680       NaN
2013-11-30 -0.138855       NaN
2013-12-31  0.029166       NaN

### Pretty Printing

In [72]:
# Pretty printing the table, to omit missing values
table = pd.pivot_table(df, index=['A', 'B'], columns=['C'])
print(table.to_string(na_rep=''))

                D                   E          
C             bar       foo       bar       foo
A     B                                        
one   A -0.233786 -0.730690  0.778850 -0.222455
      B -0.146611  0.729050 -0.319515 -0.185373
      C  0.391140  0.481786  0.402763  1.013150
three A  0.894316            0.953859          
      B           -0.281743            0.019289
      C  1.220953           -0.211392          
two   A           -0.176296            0.573914
      B -0.188322            0.704211          
      C            0.916646            0.269991


### Add Margins

Set `margins=True`. This will augment the output with additional `All` columns and `All` rows (with partial group aggregates across the categories on the columns and rows)

In [11]:
# Without Margin
df.pivot_table(index=['A', 'B'], columns='C')

D                   E          
C             bar       foo       bar       foo
A     B                                        
one   A  0.094680  0.593650 -0.361031  0.596652
      B  0.497550  0.248099 -0.034399  0.501895
      C -0.269192  0.425507  0.742233 -0.871049
three A  0.321587       NaN -0.305280       NaN
      B       NaN -0.263535       NaN -0.141598
      C  0.029166       NaN -0.544625       NaN
two   A       NaN  0.108155       NaN -0.272429
      B -0.138855       NaN  0.599492       NaN
      C       NaN  0.505092       NaN  0.661298

In [12]:
# With Margins
df.pivot_table(index=['A', 'B'], columns='C', margins=True)

D                             E                    
C             bar       foo       All       bar       foo       All
A     B                                                            
one   A  0.094680  0.593650  0.344165 -0.361031  0.596652  0.117811
      B  0.497550  0.248099  0.372825 -0.034399  0.501895  0.233748
      C -0.269192  0.425507  0.078157  0.742233 -0.871049 -0.064408
three A  0.321587       NaN  0.321587 -0.305280       NaN -0.305280
      B       NaN -0.263535 -0.263535       NaN -0.141598 -0.141598
      C  0.029166       NaN  0.029166 -0.544625       NaN -0.544625
two   A       NaN  0.108155  0.108155       NaN -0.272429 -0.272429
      B -0.138855       NaN -0.138855  0.599492       NaN  0.599492
      C       NaN  0.505092  0.505092       NaN  0.661298  0.661298
All      0.089156  0.269495  0.179325  0.016065  0.079128  0.047597

### Mimicking Cross Tabulations

We can compute the co-occurrence of an index and a column with `pivot_table`.

In [13]:
pd.pivot_table(df, index='A', columns='B', fill_value=0, aggfunc='size')

B      A  B  C
A             
one    4  4  4
three  2  2  2
two    2  2  2

Cross tabulations can be directly done using the `crosstab()` function.

In [14]:
pd.crosstab(df['A'], df['B'])

B      A  B  C
A             
one    4  4  4
three  2  2  2
two    2  2  2

In [15]:
# Normalize the output
# Sum total of all the elements in the output will be 1
pd.crosstab(df['A'], df['B'], margins=True, normalize=True)

B             A         B         C   All
A                                        
one    0.166667  0.166667  0.166667  0.50
three  0.083333  0.083333  0.083333  0.25
two    0.083333  0.083333  0.083333  0.25
All    0.333333  0.333333  0.333333  1.00

In [85]:
# Normalize ONLY the columns
# Sum of ONLY columns will be 1
pd.crosstab(df['A'], df['B'], margins=True, normalize='columns')

B         A     B     C   All
A                            
one    0.50  0.50  0.50  0.50
three  0.25  0.25  0.25  0.25
two    0.25  0.25  0.25  0.25

In [84]:
# Normalize ONLY the columns
# Sum of ONLY columns will be 1
pd.crosstab(df['A'], df['B'],margins=True, normalize='index')

B             A         B         C
A                                  
one    0.333333  0.333333  0.333333
three  0.333333  0.333333  0.333333
two    0.333333  0.333333  0.333333
All    0.333333  0.333333  0.333333

## Tiling

The `cut()` function computes groupings for the values of the input array and is often used to transform continuous variables to discrete or categorical variables

In [18]:
ages = np.array([10, 15, 13, 12, 23, 25, 28, 59, 60])
c = pd.cut(ages, bins=5)
print(c)

[(9.95, 20.0], (9.95, 20.0], (9.95, 20.0], (9.95, 20.0], (20.0, 30.0], (20.0, 30.0], (20.0, 30.0], (50.0, 60.0], (50.0, 60.0]]
Categories (5, interval[float64]): [(9.95, 20.0] < (20.0, 30.0] < (30.0, 40.0] < (40.0, 50.0] < (50.0, 60.0]]


In the above example, 3 equal-width bins are created with intervals `[(9.95, 26.667] < (26.667, 43.333] < (43.333, 60.0]`. Each element of the array is mapped to one of the bins.

If the bins keyword is an integer, then equal-width bins are formed. Alternatively we can specify custom bin-edges:

In [19]:
# Specify custom bin-edges
c = pd.cut(ages, bins=[0, 18, 35, 70])
print(c)

[(0, 18], (0, 18], (0, 18], (0, 18], (18, 35], (18, 35], (18, 35], (35, 70], (35, 70]]
Categories (3, interval[int64]): [(0, 18] < (18, 35] < (35, 70]]


In [20]:
# One-hot representation of category variables
pd.get_dummies(c)

(0, 18]  (18, 35]  (35, 70]
0        1         0         0
1        1         0         0
2        1         0         0
3        1         0         0
4        0         1         0
5        0         1         0
6        0         1         0
7        0         0         1
8        0         0         1

## Vectorization: String Operations

We have seen how vectorization helps us apply same operation to array elements (e.g. elementwise addition, multiplication, etc.).

Pandas provides rich support for vectorized string operations. These operations are also capable of handling missing/NaN values.

In [21]:
import pandas as pd
import numpy as np

In [22]:
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein', np.nan, '4325','Rohit'])
print (s)

0        Jack
1    Tom Rick
2        John
3    Einstein
4         NaN
5        4325
6       Rohit
dtype: object


### `lower()`, `upper()` and `swapcase()`

`lower()` converts strings in the Series/Index to lower case.

In [23]:
# Converts strings in the Series/Index to lower case
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein', np.nan, '4325','Rohit'])
print(s.str.lower())

0        jack
1    tom rick
2        john
3    einstein
4         NaN
5        4325
6       rohit
dtype: object


In [24]:
# Similarly, the upper function, converts Series/Index to uppercase
print(s.str.upper())

0        JACK
1    TOM RICK
2        JOHN
3    EINSTEIN
4         NaN
5        4325
6       ROHIT
dtype: object


In [25]:
# Swap case lower to upper and vice-versa
print(s.str.swapcase())

0        jACK
1    tOM rICK
2        jOHN
3    eINSTEIN
4         NaN
5        4325
6       rOHIT
dtype: object


### `len()`

In [26]:
# Compute string length
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein', np.nan, '4325','Rohit'])
print (s.str.len())

0    4.0
1    8.0
2    4.0
3    8.0
4    NaN
5    4.0
6    5.0
dtype: float64


### `strip()`

In [171]:
# Strip whitespace(including newline) from each string in the Series/index from both the sides.
s = pd.Series([' Jack ', ' Tom Rick', 'J ohn', 'Einstein '])
print(s)
print("After Stripping:")
print(s.str.strip())

0        Jack 
1     Tom Rick
2        J ohn
3    Einstein 
dtype: object
After Stripping:
0        Jack
1    Tom Rick
2       J ohn
3    Einstein
dtype: object


### `split(pattern)`

In [27]:
# Splits each string with the given pattern
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])
print(s)
print("Split Pattern:")
print(s.str.split(' '))

0        Jack
1    Tom Rick
2        John
3    Einstein
dtype: object
Split Pattern:
0         [Jack]
1    [Tom, Rick]
2         [John]
3     [Einstein]
dtype: object


### `cat(sep=pattern)`

In [28]:
# Concatenates the series/index elements with the given separator.
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])
print(s.str.cat(sep='_'))

Jack_Tom Rick_John_Einstein


### `get_dummies()`

In [29]:
# Returns the DataFrame with One-Hot Encoded values
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])

print (s.str.get_dummies())

   Einstein  Jack  John  Tom Rick
0         0     1     0         0
1         0     0     0         1
2         0     0     1         0
3         1     0     0         0


### `contains()`

In [30]:
# Returns a Boolean value True for each element if the substring contains in the element, else False.
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])
print (s.str.contains(' '))

0    False
1     True
2    False
3    False
dtype: bool


### `replace(a, b)`

In [31]:
# Replaces the value a with the value b
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])
print(s)
print("After replacing space with $:")
print(s.str.replace(' ','$'))

0        Jack
1    Tom Rick
2        John
3    Einstein
dtype: object
After replacing space with $:
0        Jack
1    Tom$Rick
2        John
3    Einstein
dtype: object


### `repeat(value)`

Repeats each element with specified number of times.

In [32]:
# Repeats each element specified number of times
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])
print (s.str.repeat(2))

0            JackJack
1    Tom RickTom Rick
2            JohnJohn
3    EinsteinEinstein
dtype: object


### `count(pattern)`

Returns count of appearance of pattern in each element.

In [33]:
# Returns count of appearance of pattern in each element. 
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])
print ("The number of 'n's in each string:")
print (s.str.count('n'))

The number of 'n's in each string:
0    0
1    0
2    1
3    2
dtype: int64


### `startswith(pattern)` and `endswith(pattern)`

In [35]:
# Returns true if the element in the Series/Index starts with the given pattern.
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])
print ("Strings that start with 'J':")
print (s.str. startswith ('J'))

Strings that start with 'J':
0     True
1    False
2     True
3    False
dtype: bool


In [36]:
# Returns true if the element in the Series/Index ends with the pattern.
print ("Strings that end with 'n':")
print (s.str.endswith('n'))

Strings that end with 'n':
0    False
1    False
2     True
3     True
dtype: bool


### `find(pattern)` and `findall(pattern)`

In [37]:
# Returns the first position of the first occurrence of the pattern
s = pd.Series(['Jack', 'Tom Rick', 'John', 'Einstein'])
print (s.str.find('a'))

0    1
1   -1
2   -1
3   -1
dtype: int64


In [38]:
# Returns a list of all occurrences of the pattern.
print (s.str.findall('i'))

0        []
1       [i]
2        []
3    [i, i]
dtype: object


### `islower()` and `isupper()`

In [39]:
# Checks whether all characters in each string in the Series/Index in lower case or not. Returns Boolean
s = pd.Series(['JACK', 'Tom Rick', 'john', 'Einstein'])
print (s.str.islower())

0    False
1    False
2     True
3    False
dtype: bool


In [40]:
# Checks whether all characters in each string in the Series/Index in upper case or not. Returns Boolean.
print (s.str.isupper())

0     True
1    False
2    False
3    False
dtype: bool


### `isnumeric()`

Checks whether all characters in each string in the Series/Index are numeric. Returns Boolean.

In [6]:
# Checks whether all characters in each string in the Series/Index are numeric. Returns Boolean.
s = pd.Series(['Kate', 'Johnny', '2456', '123n'])
print (s.str.isnumeric())
print (s.str.isalnum())
print (s.str.isalpha())

0    False
1    False
2     True
3    False
dtype: bool
0    True
1    True
2    True
3    True
dtype: bool
0     True
1     True
2    False
3    False
dtype: bool


## High Performance Pandas

Pandas are used to process large DataFrames. As such pandas is very efficient in performing vectorized operations (as it pushes the basic operations to C language).

However, intermediate computations are explicitly stored in memory as temporary objects. Thus making computations slower specially when we have large datasets to process.

Pandas provides `eval()` and `query()` tools for efficient operations.

### `eval()` Function

Uses string expressions to efficiently compute operations using DataFrame.

In [9]:
import pandas as pd
import numpy as np

nrows, ncols = 100000, 100
rng = np.random.RandomState(42)
df1, df2, df3, df4 = (pd.DataFrame(rng.rand(nrows, ncols))
                      for i in range(4))
print(rng)
print("df1-"*60)
print(df1)
print("df2-"*60)
print(df2)
print("df3-"*60)
print(df3)
print("df4-"*60)
print(df4)
print("-"*60)
# Compute sum of four dataframes -- traditional way
%timeit df1 + df2 + df3 + df4

# Compute the sum with pd.eval
%timeit pd.eval('df1 + df2 + df3 + df4')    # Operations represented as a string

RandomState(MT19937)
df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-df1-
             0         1         2         3         4         5         6   \
0      0.374540  0.950714  0.731994  0.598658  0.156019  0.155995  0.058084   
1      0.031429  0.636410  0.314356  0.508571  0.907566  0.249292  0.410383   
2      0.642032  0.084140  0.161629  0.898554  0.606429  0.009197  0.101472   
3      0.051682  0.531355  0.540635  0.637430  0.726091  0.975852  0.516300   
4      0.103124  0.902553  0.505252  0.826457  0.320050  0.895523  0.389202   
...         ...       ...       ...       ...       ...       ...       ...   
99995  0.071979  0.439323  0.188588  0.586705  0.640611  0.662409  0.318503   
99996  0.313411  0.010490  0.469216  0.600825  0.451085  0.496918  0.983128   
99997  0.560873  0.647396  

121 ms ± 3.83 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
53.8 ms ± 3.53 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


The `eval()` function supports Arithmetic, Comparison, Bitwise, Object Attributes and Indices based operations.

### `query()` Function

In [15]:
df = pd.DataFrame(rng.randint(0, 1000, (10000, 3)), columns=['A', 'B', 'C'])
print(df)
# logical and relational expression
result1 = df[(df.A < 100) & (df.B < 100)]
print(result1)
# using query - specify the expression as a string
result2 = df.query('A < 100 and B < 100')

np.allclose(result1, result2)

        A    B    C
0     746  705  863
1     291  115  654
2     259  669  283
3     929  379  179
4     589  221  677
...   ...  ...  ...
9995  464   24  190
9996   36  801  909
9997  453  211  595
9998  661  707  940
9999  460  959  623

[10000 rows x 3 columns]
       A   B    C
6     63  51  706
117   55   9  144
135   14  90  291
222   72  72  846
232    0  76   99
...   ..  ..  ...
9239   8  29  676
9559  10  34  372
9588  99  77   11
9645  92  50  475
9739  17  35   46

[98 rows x 3 columns]


True

`allclose()` helps us ensure that `result1` and `result2` are identical within tolerance limits.

**NOTE:** The benefits of `eval()` and `query()` are with respect to reduction in memory usage and time saved during computation. However, these benefits, in general, are realised when we have large arrays or datasets to work with.

## Categorical Data

Categorical variables can take on only a limited, and usually fixed number of possible values. Besides the fixed length, categorical data might have an order but cannot perform numerical operation. Categorical are a Pandas data type.

The categorical data type is useful in the following cases,

- A string variable consisting of only a few different values, e.g. Male or Female. Converting such a string variable to a categorical variable will save some memory.

- The lexical order of a variable is not the same as the logical order (“one”, “two”, “three”). By converting to a categorical and specifying an order on the categories, sorting and min/max will use the logical order instead of the lexical order.

- As a signal to other python libraries that this column should be treated as a categorical variable (e.g. to use suitable statistical methods or plot types).

### Create Categorical Data (Nominal and Ordinal)

In [188]:
import pandas as pd

# Specify the dtype as "category" in pandas object creation
s = pd.Series(["a", "b", "c", "a"], dtype="category")
print(s)

0    a
1    b
2    c
3    a
dtype: category
Categories (3, object): [a, b, c]


While we have passed four elements to the series object, the number of categories are only three. These are also referred to as Nominal (name) categories.

In [189]:
# Using the standard pandas Categorical constructor
cat = pd.Categorical(['a', 'b', 'c', 'a', 'b', 'c'])
print(cat)

[a, b, c, a, b, c]
Categories (3, object): [a, b, c]


In [190]:
# What if the data doesn't have some categories that we need
cat = pd.Categorical(['a', 'b', 'c', 'a', 'b', 'c', 'd'], ['c', 'b', 'a', 'e'])
print(cat)

[a, b, c, a, b, c, NaN]
Categories (4, object): [c, b, a, e]


Here, the second argument signifies the categories. Thus, any value which is not present in the categories will be treated as NaN.

Now, take a look at the following example −

In [16]:
# If we need logical ordering of categories.
# Lets assume we have fever values for 5 people, as Very High, High, and Normal
s = pd.Categorical(['Very High', 'High', 'Normal', 'Very High', 'Normal'], ['Normal', 'Very High', 'High'], ordered=True)
print(s)

[Very High, High, Normal, Very High, Normal]
Categories (3, object): [Normal < Very High < High]


In [17]:
print(s.categories)
print(s.ordered)

Index(['Normal', 'Very High', 'High'], dtype='object')
True


In [193]:
s = s.add_categories(['Fatal'])
print(s)

[Very High, High, Normal, Very High, Normal]
Categories (4, object): [Normal < High < Very High < Fatal]


In [194]:
s = s.remove_categories(['Fatal'])
print(s)

[Very High, High, Normal, Very High, Normal]
Categories (3, object): [Normal < High < Very High]


In [195]:
# Compare categorical data. Similarly we can do comparisons of ordered category.
print(s[0] < s[1])

False


## Regular Expressions

A regular expression--also known as RE, regex or regexp--is a sequence of characters that define a search pattern. Usually such patterns are used by string searching algorithms for "find" or "find and replace" operations on strings, or for input validation. [Source: Wikipedia]

The Python module `re` provides full support for Perl-like regular expressions in Python. The `re` module raises the exception re.error if an error occurs while compiling or using a regular expression. We will briefly cover the two important functions widely used with regex.

In [18]:
import re

We would cover two important functions, which would be used to handle regular expressions. But a small thing first: There are various characters, which would have special meaning when they are used in regular expression. To avoid any confusion while dealing with regular expressions, we would use Raw Strings as r'expression'.

### `match()` Function
This function attempts to match regex pattern to the string.

In [197]:
# Syntax
# re.match(pattern, string, flags=0)

In [198]:
help(re.match)

Help on function match in module re:

match(pattern, string, flags=0)
    Try to apply the pattern at the start of the string, returning
    a Match object, or None if no match was found.



`re.match()` returns a match object (at the start of the string) on success, None on failure. `group(num)` or `groups()` returns the string matched by the RE.

### RE: Match Example

In [19]:
import re

string = "02/02/2020 was an important day"
pattern = "\d\d/\d\d/\d\d\d\d"    # \d - Matches any digit
matchObj = re.match(pattern, string)
if matchObj:
    print ("matchObj.group() : ", matchObj.group())
else:
    print ("No match!!")

#Try to create an example to serach for PAN number at the start of a string

matchObj.group() :  02/02/2020


### `search()` Function

This function searches for first occurrence of RE pattern within string with optional flags.

In [200]:
# Syntax
# re.search(pattern, string, flags=0)

In [201]:
help(re.search)

Help on function search in module re:

search(pattern, string, flags=0)
    Scan through string looking for a match to the pattern, returning
    a Match object, or None if no match was found.



### RE: Search Example

In [202]:
string = "This 02/02/2020 was an important day"
pattern = "\d\d/\d\d/\d\d\d\d"    # \d - Matches any digit

print('Let''s try matching')
matchObj = re.match(pattern, string)
if matchObj:
    print ("matchObj.group() : ", matchObj.group())
else:
    print ("No match!!")
    
# match() doesn't find the date, let's use search
print('Let''s try searching')
matchObj = re.search(pattern, string)
if matchObj:
    print ("matchObj.group() : ", matchObj.group())
else:
    print ("No match!!")

Lets try matching
No match!!
Lets try searching
matchObj.group() :  02/02/2020


### Search and Replace

In [203]:
# Syntax
# re.sub(pattern, repl, string, max=0)

In [204]:
help(re.sub)

Help on function sub in module re:

sub(pattern, repl, string, count=0, flags=0)
    Return the string obtained by replacing the leftmost
    non-overlapping occurrences of the pattern in string by the
    replacement repl.  repl can be either a string or a callable;
    if a string, backslash escapes in it are processed.  If it is
    a callable, it's passed the Match object and must return
    a replacement string to be used.



### RE: Search and Replace Example

In [30]:
import re

phone = "2004-959-559 # This is Phone Number"

# Delete Python-style comments
# Delete python comments starting with # followed by zero or more characters (.*) till the end of the line ($)
num = re.sub(r'#.*$', "", phone)
print ("Phone Num : ", num)
num1 = re.match(r'[\d]+-[\d]+-[\d]+', phone)
print(num1)

# Remove anything other than digits
num = re.sub(r'\D', "", phone)    
print ("Phone Num : ", num)

Phone Num :  2004-959-559 
<re.Match object; span=(0, 12), match='2004-959-559'>
Phone Num :  2004959559


### Regular Expression Modifiers: Option Flags

Regular expression literals may include an optional modifier to control various aspects of matching. The modifiers are specified as an optional flag. You can provide multiple modifiers using exclusive OR (|), as shown previously and may be represented by one of these −

- re.I: 
  Performs case-insensitive matching.
- re.L: 
  Interprets words according to the current locale. This interpretation affects the alphabetic group (\w and \W), as well as word boundary behavior(\b and \B).
- re.M: 
  Makes ($) match the end of a line (not just the end of the string) and makes ^ match the start of any line (not just the start of the string).
- re.S
  Makes a period (dot) match any character, including a newline.
- re.U
  Interprets letters according to the Unicode character set. This flag affects the behavior of \w, \W, \b, \B.
- re.X
  Permits "cuter" regular expression syntax. It ignores whitespace (except inside a set [] or when escaped by a backslash) and treats unescaped # as a comment marker.



### Regular Expression Patterns

Except for control (meta) characters, (+ ? . * ^ $ ( ) [ ] { } | \), all characters match themselves. To include a control character in a search expression we need to escape the control character by preceding it with a backslash.

- \^ Matches beginning of a line
- \+ One or more occurrences of the preceding letter
- \* Zero or more occurrences of the preceding letter
- \. Any letter <br>
...


RE are very powerful and provide a lot of features. We encourage you to explore them in more detail by referring to https://docs.python.org/3/howto/regex.html

## Practice Questions

- Write a Pandas program to create a Pivot table with multiple indexes from a given excel sheet (Salesdata.xlsx). 
- Write a Python program to check that a string contains only a certain set of characters (in this case a-z, A-Z and 0-9).
- Write a Python program that matches a string that has an a followed by zero or one 'b'.
- Write a Python program to find sequences of lowercase letters joined with a underscore.
- Write a Python program that matches a word at the end of a string, with optional punctuation.
- Write a Python program where a string will start with a specific number.

In [51]:
df = pd.read_excel('Salesdata.xlsx')
print(type(df.loc[0,'OrderDate']),
      type(df.loc[0,'Region']),
      type(df.loc[0,'Units']),
      type(df.loc[0,'Sale_amt']))
df.head()

<class 'pandas._libs.tslibs.timestamps.Timestamp'> <class 'str'> <class 'numpy.int64'> <class 'numpy.float64'>


OrderDate   Region  Manager   SalesMan          Item  Units  Unit_price  \
0 2018-01-06     East   Martha  Alexander    Television     95      1198.0   
1 2018-01-23  Central  Hermann     Shelli  Home Theater     50       500.0   
2 2018-02-09  Central  Hermann       Luis    Television     36      1198.0   
3 2018-02-26  Central  Timothy      David    Cell Phone     27       225.0   
4 2018-03-15     West  Timothy    Stephen    Television     56      1198.0   

   Sale_amt  
0  113810.0  
1   25000.0  
2   43128.0  
3    6075.0  
4   67088.0

In [43]:
pattern = r'\d+'
for col in df:
    for row in col:
        re.match(pattern, row)
        
        

SyntaxError: unexpected EOF while parsing (<ipython-input-43-a2a4f958f0bd>, line 4)

## Source(s):
- https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html
- https://www.tutorialspoint.com/python/python_reg_expressions.htm
- https://docs.python.org/3/howto/regex.html